File: client/src/app/services/time/time.service.ts
import { Injectable } from '@angular/core';
import { SocketHandlerService } from '@app/services/socket-handler/socket-handler.service';
import { MULTIPLICATION_FACTOR, PANIC_ALERT_DELAY } from '@common/constants/match-constants';
import { TimerEvents } from '@common/events/timer.events';
import { TimerInfo } from '@common/interfaces/timer-info';

@Injectable({
    providedIn: 'root',
})
export class TimeService {
    isTimerPaused: boolean;
    isPanicModeDisabled: boolean;
    isPanicking: boolean;
    isAlertDisplayed: boolean;
    alertSymbol: string;
    private counter: number;
    private initialValue: number;

    constructor(private readonly socketService: SocketHandlerService) {
        this.counter = 0;
        this.initialValue = 0;
        this.isPanicModeDisabled = false;
        this.isPanicking = false;
        this.isAlertDisplayed = false;
        this.isTimerPaused = false;
    }

    get time() {
        return this.counter;
    }

    get duration() {
        return this.initialValue;
    }

    set time(newTime: number) {
        this.counter = newTime;
    }

    listenToTimerEvents() {
        this.handleTimer();
        this.handleStopTimer();
        this.onPanicTimer();
        this.onPauseTimer();
        this.onResumeTimer();
        this.onDisablePanicTimer();
    }

    startTimer(roomCode: string, time: number): void {
        this.socketService.send(TimerEvents.StartTimer, { roomCode, time });
    }

    stopTimer(roomCode: string): void {
        this.socketService.send(TimerEvents.StopTimer, { roomCode });
    }

    pauseTimer(roomCode: string): void {
        this.socketService.send(TimerEvents.PauseTimer, roomCode);
    }

    triggerPanicTimer(roomCode: string): void {
        this.isPanicking = true;
        this.socketService.send(TimerEvents.PanicTimer, roomCode);
    }

    handleTimer(): void {
        this.socketService.on(TimerEvents.Timer, (timerInfo: TimerInfo) => {
            this.counter = timerInfo.currentTime;
            this.initialValue = timerInfo.duration;
        });
    }

    handleStopTimer(): void {
        this.socketService.on(TimerEvents.StopTimer, () => {
            this.isPanicking = false;
        });
    }

    onPanicTimer() {
        this.socketService.on(TimerEvents.PanicTimer, () => {
            this.alertSymbol = '❗';
            this.displayPanicAlert();
        });
    }

    onPauseTimer() {
        this.socketService.on(TimerEvents.PauseTimer, () => {
            this.alertSymbol = 'II';
            this.displayPanicAlert();
        });
    }

    onResumeTimer() {
        this.socketService.on(TimerEvents.ResumeTimer, () => {
            this.alertSymbol = '▶';
            this.displayPanicAlert();
        });
    }

    onDisablePanicTimer(): void {
        this.socketService.on(TimerEvents.DisablePanicTimer, () => {
            this.isPanicModeDisabled = true;
        });
    }

    computeTimerProgress(): number {
        return (this.time / this.duration) * MULTIPLICATION_FACTOR;
    }

    togglePauseStatus() {
        this.isTimerPaused = !this.isTimerPaused;
    }

    private displayPanicAlert() {
        this.isAlertDisplayed = true;
        setTimeout(() => {
            this.isAlertDisplayed = false;
        }, PANIC_ALERT_DELAY);
    }
}

File: client-mobile/app/src/main/java/com/example/polyquiz/classicmode/domain/TimeService.kt
package com.example.polyquiz.classicmode.domain

import com.example.vanillaprototype.socket.SocketHandler
import com.example.polyquiz.constants.TimerEvents
import com.example.polyquiz.constants.TimerInfo
import com.google.gson.Gson

object TimeService {
    var isTimerPaused: Boolean = false
    //var isPanicModeDisabled: Boolean = false
    //var isPanicking: Boolean = false
    //var isAlertDisplayed: Boolean = false
   // var alertSymbol: String = ""
    private var counter: Int = 0
    private var initialValue: Int = 0

    private val mSocket = SocketHandler.getSocket()

    val time: Int
        get() = counter

    val duration: Int
        get() = initialValue

    var timeSetter: Int
        get() = counter
        set(newTime) {
            counter = newTime
        }

    fun handleTimer() {
        mSocket.on(TimerEvents.TIMER.value) { args ->
            if (args.isNotEmpty() && args[0] != null) {
                val timerInfo = Gson().fromJson(args[0].toString(), TimerInfo::class.java)
                counter = timerInfo.currentTime
                initialValue = timerInfo.duration
            }
        }
    }
    fun startTimer(roomCode: String, time: Int) {
        val timeCodeObject = mapOf("roomCode" to roomCode, "time" to time)
        val timeJsonObject = Gson().toJson(timeCodeObject)
        mSocket.emit(TimerEvents.START_TIMER.value, timeJsonObject);
    }
    fun stopTimer(roomCode: String) {
        val roomCodeObject = mapOf("roomCode" to roomCode)
        val roomJsonObject = Gson().toJson(roomCodeObject)
        mSocket.emit(TimerEvents.STOP_TIMER.value, roomJsonObject);
    }

    fun computeTimerProgress(): Float {
        val progress = (counter.toFloat() / duration.toFloat()) * 100
        return progress
    }

//    fun joinRoom(roomCode: String, username: String) {
//        val roomusernameObject = mapOf("roomCode" to roomCode, "username" to username)
//        val roomCodeJsonObject = Gson().toJson(roomusernameObject)
//        mSocket.emit("joinRoom", roomCodeJsonObject)
//        val roomCodeObject = mapOf("roomCode" to roomCode)
//        val roomJsonObject = Gson().toJson(roomCodeObject)
//        mSocket.emit("sendPlayersData", roomJsonObject)
//    }
}
